File: Data Structures & Algorithms/house-robber/submission-8.java
class Solution {
    public int rob(int[] nums) {
        return rec(0, nums);
    }
    public int rec(int n, int[] nums) {
        if (n >= nums.length){
            return 0;
        }
        return Math.max(nums[n] + rec(n + 2, nums), rec(n + 1, nums));
    }
}

File: Data Structures & Algorithms/house-robber-ii/submission-8.cpp
class Solution {
    unordered_map<int, int> cache;
public:
    int rob(vector<int>& nums) {
        int n{static_cast<int>(nums.size())};
        auto back = [&](auto self, int i, int end){
            if (i >= end) {
                return 0;
            }
            return max(self(self, i + 1, end), self(self, i + 2, end) + nums[i]);
        };
        return max(back(back, 0, n - 1), back(back, 1, n));
    }
};
